src/lib/measure.c

```diff
@@ -0,0 +1,150 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>
+ */
+
+#include <sof/measure.h>
+#include <sof/timer.h>
+
+struct measure_context measure_init(uint32_t average_steps)
+{
+	struct measure_context context;
+
+	memset_s(&context, sizeof(struct measure_context), 0,
+		 sizeof(struct measure_context));
+
+	context.steps = average_steps;
+	context.samples =
+		(struct measure_sample *)rzalloc(RZONE_RUNTIME,
+						 SOF_MEM_CAPS_RAM,
+						 average_steps *
+						 sizeof(struct measure_sample));
+
+	return context;
+}
+
+void measure_free(struct measure_context *context)
+{
+	rfree(context->samples);
+	context->samples = NULL;
+}
+
+static inline void measure_update_min_stat(struct measure_context *context)
+{
+	struct measure_sample *sample = &context->samples[context->cur_step];
+
+	if (sample->time_delta <
+	    context->stats[MEASURE_STAT_MIN].time_delta)
+		context->stats[MEASURE_STAT_MIN].time_delta =
+			sample->time_delta;
+
+	if (sample->ccount_delta <
+	    context->stats[MEASURE_STAT_MIN].ccount_delta)
+		context->stats[MEASURE_STAT_MIN].ccount_delta =
+			sample->ccount_delta;
+}
+
+static inline void measure_update_max_stat(struct measure_context *context)
+{
+	struct measure_sample *sample = &context->samples[context->cur_step];
+
+	if (sample->time_delta >
+	    context->stats[MEASURE_STAT_MAX].time_delta)
+		context->stats[MEASURE_STAT_MAX].time_delta =
+			sample->time_delta;
+
+	if (sample->ccount_delta >
+	    context->stats[MEASURE_STAT_MAX].ccount_delta)
+		context->stats[MEASURE_STAT_MAX].ccount_delta =
+			sample->ccount_delta;
+}
+
+static inline void measure_update_avg_stat(struct measure_context *context)
+{
+	int avg_step;
+
+	context->stats[MEASURE_STAT_AVG].ccount_delta = 0;
+	context->stats[MEASURE_STAT_AVG].time_delta = 0;
+
+	for (avg_step = 0; avg_step < context->steps; ++avg_step) {
+		context->stats[MEASURE_STAT_AVG].ccount_delta +=
+			context->samples[avg_step].ccount_delta;
+		context->stats[MEASURE_STAT_AVG].time_delta +=
+			context->samples[avg_step].time_delta;
+	}
+	context->stats[MEASURE_STAT_AVG].ccount_delta /=
+		context->steps;
+	context->stats[MEASURE_STAT_AVG].time_delta /=
+		context->steps;
```

Comment:
If the sampling size was a power of 2 then the expensive divisions here and within this file in general could be a cheap shift.

Reply:
Good idea

Reply:
Actually, in order to find out the shift I'd need to log2 the amount of steps. Would that really be faster?